collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryQueue.kt

```diff
@@ -0,0 +1,84 @@
+package com.bitmovin.analytics.retryBackend
+
+import android.util.Log
+import java.util.Calendar
+import java.util.Date
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.Comparator
+import kotlin.math.pow
+
+class RetryQueue {
+    private val TAG = "RetryQueue"
+    private val lock = ReentrantLock()
+    private val MAX_RETRY_TIME = 300 // in seconds
+    private val MAX_RETRY_SAMPLES = 100
+    private val MAX_BACKOFF_INTERVAL = 64
+
+    fun getMaxSampleNumber() = MAX_RETRY_SAMPLES
+    fun now() = Date()
+
+    private var retrySamplesList = mutableListOf<RetrySample<Any>>()
+    private val sampleComparator = Comparator<RetrySample<Any>> { a, b ->
+        when {
+            (a.scheduledTime == b.scheduledTime) -> 0
+            (a.scheduledTime.before(b.scheduledTime)) -> -1
+            else -> 1
+        }
+    }
+
+    fun addSample(retrySample: RetrySample<Any>) {
+        try {
+            lock.lock()
+            retrySample.retry++
+
+            val backOffTime = minOf(2.toDouble().pow(retrySample.retry).toInt(), MAX_BACKOFF_INTERVAL)
+            retrySample.totalTime += backOffTime
+
+            // more than 5min in queue
+            if (retrySample.totalTime > MAX_RETRY_TIME) {
+                return
+            }
+
+            retrySample.scheduledTime = Calendar.getInstance().run {
+                add(Calendar.SECOND, backOffTime)
+                time
+            }
+            if (retrySamplesList.size >= getMaxSampleNumber()) {
+                val removeSample = retrySamplesList.last()
+                retrySamplesList.remove(removeSample)
+                Log.d(TAG, "removed sample")
```

Comment:
more extensive logging would be helpful (`removed oldest sample due to queue being over capacity of %d` or something)